### PR TITLE
New MSD and residence time modules for MDAnalysis

### PR DESCRIPTION
**Pull request by abiedermann**
Hello!

MDAnalysis has been very helpful for my research, so I would like to share a couple of analysis tools which might be useful to other users. The first is a faster and more specific script for calculating mean squared displacements and the other implements an extension of the stable-states residence time method described by Laage and Hynes. The modules scale well with simulation time and length, as I have used them to analyze 100 ns simulations of ~140,000 atoms. Additionally, the MSD script has been validated against the gromacs gmx msd module.

Attached is a document which explains the purpose of the scripts in greater detail. Please let me know if you think these tools would be helpful to other users as well as if you would like me to make additional modifications to the code.

Best,
Andrew

[StateDependentAnalysis.pdf](https://github.com/MDAnalysis/mdanalysis/files/530648/StateDependentAnalysis.pdf)

### CODE REVIEW

**Comment by tylerjereddy** [comment]
I haven't had a chance to properly look things over yet, but note that the test data file sizes are almost certainly too large at the moment:

- `BPSH_50ps.xtc` : 12.9 MB
-  `BPSH_50ps.tpr`: 7.29 MB

**Comment by richardjgowers** [comment]
Hey Andrew. Looks good overall, there'll be a few changes needed here and there. 

To shrink the test files, you could reduce the system to only 10 atoms. The tests don't need to produce useful data, just let us know if it ever gets broken.

File: package/MDAnalysis/analysis/MSD.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+
+        # pre-allocate position array
+        pos = [[np.array([]) for j in range(self.len_msd)]
+               for i in range(n_sel)]
+
+        print("Populating deque...")
+        for ts in self.universe.trajectory[self.t0:self.t0
+                                           + self.len_msd]:
+            this_frame = ts.frame
+            selections = self._select(this_frame)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: coordinates of selections are automatically updated. You can just do this at the beginning of the loop. In fact you can to this at the beginning of the function and then set `n_sel = len(selections)`.

File: package/MDAnalysis/analysis/MSD.py [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+
+            for i in range(n_sel):
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: Rather loop over elements in `selections` with `for sel in selections`.

File: package/MDAnalysis/analysis/MSD.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+                if selections[i] is not None:  # if there is data
+                    pos[i][this_frame-self.t0] = selections[i].positions
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: To be on the save side you should copy the positions instead of just using a reference. Just add `.copy()` at the end of the line.

File: package/MDAnalysis/analysis/MSD.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+                                    ) / (n_samples[i][ts-j]+1)
+                    n_samples[i][ts-j] += 1
+            
+            if j % 100 == 0:
+                print("Frame: "+str(j))
+                if self.write_output:
+                    pickle.dump([msd, n_samples], open(self.out_name, 'wb'))
+
+            # Update deques
+            self._update_deques(pos, set_list, dict_list, j)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: Why do you need to update the deques?

File: package/MDAnalysis/analysis/MSD.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+
+        n_samples = np.zeros((n_sel, self.len_msd), dtype='int')
+        msd = np.zeros((n_sel, self.len_msd), dtype='float32')
+        
+        calc_cutoff = self.tf
+        if self.max_data:
+            calc_cutoff = len(self.universe.trajectory) - 1
+
+        # for each restart point
+        for j in range(self.t0, self.tf+1, self.dt_restart):
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: So `dt_restart` is the steps between frames that are analyzed for the MSD. Then please call it `step` to be consistent with the default python/MDAnalysis terminology.

File: package/MDAnalysis/analysis/MSD.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+        top_frame = this_restart + self.len_msd
+        calc_cutoff = self.tf
+        if self.max_data:
+            calc_cutoff = len(self.universe.trajectory) - 1
+
+        if top_frame+self.dt_restart > calc_cutoff:
+            # feed in zeros
+            for i in range(len(self.select_list)):
+                pos[i].append(np.array([]))
+                set_list[i].append(set())
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This isn't a good variable name. It doesn't me any clue about what is stored in that variable. Instead of stating the type in the name state the content that makes the algorithm easier to understand. I always try to pick variable names that allow me to know what the variable contains just from the name even if I worked on something completely for a year before looking at the code again.

File: package/MDAnalysis/analysis/MSD.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+
+    Parameters
+    ----------
+    universe : object
+        Universe object containing trajectory data
+    select_list : list
+        List of selection strings
+    t0 : int
+        Time when analysis begins (in frames)
+    tf : int
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: please use more meaningful variable names like `stop_frame`. The two letters `tf` can have a large number of meanings and it makes the code where you actually use the variables hard to understand.

File: package/MDAnalysis/analysis/MSD.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+                    
+                    # find mutual atoms at times 0 and ts-j
+                    shared0 = [dict_list[i][0][k] for k in atoms_in_sel]
+                    shared = [dict_list[i][ts-j][k] for k in atoms_in_sel]
+                    
+                    # move to next restart if there's nothing to evaluate
+                    if len(shared) == 0:
+                        break  # skip to next restart
+                    
+                    msd[i][ts-j] = (msd[i][ts-j]*n_samples[i][ts-j] +
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: You actually are using an inefficient algorithm here to calculate the MSD. One issue is that you don't use all available information in the trajectory and another that this has a long runtime, a FFT based algorithm will be much faster. For the runtime this version here though is good for if have some atoms that are some frame in your selected reagion and some other frames outside of it. I'll write more about the sampling issue and FFT algorithms tomorrow.

File: package/MDAnalysis/analysis/MSD.py (hunk 1 above)
Comment (abiedermann) [inline package/MDAnalysis/analysis/MSD.py]: Re-selecting at every frame makes sure the selections are up to date. The coordinates update dynamically, but the atoms included in the selection do not.

File: package/MDAnalysis/analysis/MSD.py (hunk 4 above)
Comment (abiedermann) [inline package/MDAnalysis/analysis/MSD.py]: The deques holds all of the positions, atom ids, mapping dictionaries needed for MSD calculations at the current restart point. This allows us to avoid reading trajectory frames more than once.

**Comment by abiedermann** [comment]
Thanks for the feedback! I will work on getting the code up to par.

File: package/MDAnalysis/analysis/MSD.py (hunk 8 above)
Comment (richardjgowers) [inline package/MDAnalysis/analysis/MSD.py]: This would be a nice optimisation, but I'm happy to let this PR use this method for now

File: package/MDAnalysis/analysis/MSD.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+
+import numpy as np
+import pickle
+# from six import range
+
+import MDAnalysis
+from MDAnalysis.lib.distances import squared_distance_vector
+
+
+class MSD(object):
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/MSD.py]: This would be nice to have, but I'm not sure if the access pattern that's being used works with the Base. (This pattern is similar to what @jdetle did with the RMSD analysis though right?)

File: package/MDAnalysis/analysis/MSD.py [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+"""
+MSD --- :mod:'MDAnalysis.analysis.MSD'
+==============================================================
+
+:Author: Andrew Biedermann
+:Year: 2016
+:Copyright: GNU Public License v3
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: Can you please change the license. Having mixed license code in a project is a huge pain for maintaining code.

File: package/MDAnalysis/analysis/MSD.py [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+simulation, the following preprocessing is necessary:
+    >>> gmx trjconv -f mysim.xtc -s mysim.tpr -o processed_sim.xtc -pbc nojump
+
+MSD command line usage
+----------------------
+
+This calculates the MSD of 3 atom selections over the first 50000 frames
+of the trajectory:
+
+    >>> ./MSD.py -f mysim.xtc -s mysim.tpr --sel 'name NA,name NA and prop z
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: Please remove all mentions and docs of the CLI version. We don't support scripts in MDAnalysis right now.

File: package/MDAnalysis/analysis/MSD.py [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+    The variables len_msd and dt_restart can be used to mitigate
+    computational and memory costs. This allows calculations to scale
+    well with simulation time and size--though you must have
+    enough RAM to load the trajectory during calculation.
+
+    The algorithm uses all available data in calculation, as the
+    calculation incorporates all atoms which have remained in the
+    dynamic selection over the interval, tau, from the restart.
+
+    Parameters
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This is still an issue

File: package/MDAnalysis/analysis/MSD.py [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+from __future__ import absolute_import, print_function, division
+
+import sys
+import getopt
+import warnings
+from collections import deque
+
+import numpy as np
+import pickle
+# from six import range
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This should be uncommented and be imported after `__future__` because it changes the language behavior.

File: package/MDAnalysis/analysis/MSD.py (hunk 4 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: You actually read the trajectory several times. As I can see it this follows the normal N**2 algorithm.

File: package/MDAnalysis/analysis/MSD.py [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+                    temp_list = sel.atoms.ids
+                    # store set of atom ids which satisfy selection
+                    set_list[i][this_frame-self.start_frame] = set(temp_list)
+                    # link atom id to array index
+                    for j in range(len(temp_list)):
+                        dict_list[i][this_frame-self.start_frame][temp_list[j]] = j
+                    
+            if this_frame % (self.len_msd/10) == 0:
+                print("Frame: "+str(this_frame))
+        # converting lists to deques
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: Why is the convertion to a deque necessary? You never do a pop operation on them at the start.

File: package/MDAnalysis/analysis/MSD.py (hunk 10 above)
Comment (abiedermann) [inline package/MDAnalysis/analysis/MSD.py]: Sorry, I thought 'GNU Public License v3' would be GPLv2 and higher. I will change it to 'GNU Public License v2.

File: package/MDAnalysis/analysis/MSD.py (hunk 14 above)
Comment (abiedermann) [inline package/MDAnalysis/analysis/MSD.py]: I set maxlen=self.len_msd, so any data fed in one end pops out data on the other. This way the deque only hold data relevant to calculations at the current restart.

File: package/MDAnalysis/analysis/MSD.py (hunk 4 above)
Comment (abiedermann) [inline package/MDAnalysis/analysis/MSD.py]: The algorithm's N**2, but instead of looping through trajectory frames, it's looping through data in the deques which is much faster. The algorithm gets data from the trajectory in _update_deques(), then it runs the O(n**2) operation in this data in _process_pos_data().

**Comment by orbeckst** [comment]
Regarding file sizes https://github.com/MDAnalysis/mdanalysis/pull/1034#issuecomment-253963948 : If you need a trajectory with water (TIP4P)  and protein you can use
```python
from MDAnalysis.tests.datafiles import TPR, XTC
```
but the protein is split across periodic boundaries and it's only 10 frames or so. Alternatively
```python
from MDAnalysis.tests.datafiles import PDB_sub_sol, XTC_sub_sol
```
is a protein centered in the box with water (TIP4P) and 4 frames.

If you just need water, use
```python
from MDAnalysis.tests.datafiles import waterPSF, waterDCD
```
with only 5 waters and 10 frames.

If none of these files are sufficient to test your problem then tell us what you need. For instance, we could extend the `XTC_sub_sol` trajectory – it is fairly small with ~20k atoms.

File: package/MDAnalysis/analysis/MSD.py (hunk 14 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This seems very fragile. You would need to ensure that you to exactly `len_msd` appends in `update_deque` to remove any stale data points in them. If you do to many or to little the MSD will be slightly off.  It would be more robust if you leave this a list and empty them in `update_deque` and check there that they have been filled correctly.

File: package/MDAnalysis/analysis/MSD.py (hunk 4 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: OK now I understand what is happening. You read the trajectory piece wise and calculate all MSD values on the pieces you read. I now understand also why you are using the deque. That is a smart trick. Sorry for taking so long to understand this.

File: package/MDAnalysis/lib/distances.py [hunk 15]
```diff
@@ -114,29 +114,22 @@ def _run(funcname, args=None, kwargs=None, backend="serial"):
...
-        * ``tri_vecs`` triclinic box vectors
-        * ``tri_box`` triclinic box lengths and angles
-
-    Raises
-    ------
-    TypeError
-        If box is not float32.
-    ValueError
-        If box type not detected.
+    :Arguments:
```
Comment (kain88-de) [inline package/MDAnalysis/lib/distances.py]: This looks like an error that must have happened when you merged the develop branch. If you split this PR we can get rid of this as well and it will also make it easier to remove the big test files from the history, I would remove them anyway before the merge because otherwiese they would be stuck in the history and we would have to download them everytime with a fresh checkout.

File: package/MDAnalysis/analysis/MSD.py (hunk 8 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: The FFT approach isn't generalizable to this algorithm. But this version is still smart and I suspect very fast for an N**2 algorithm.

File: package/MDAnalysis/analysis/MSD.py [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+                    shared = [dict_list[i][ts-j][k] for k in atoms_in_sel]
+                    
+                    # move to next restart if there's nothing to evaluate
+                    if len(shared) == 0:
+                        break  # skip to next restart
+                    
+                    msd[i][ts-j] = (msd[i][ts-j]*n_samples[i][ts-j] +
+                                    squared_distance_vector(pos[i][0][shared0],
+                                    pos[i][ts-j][shared]).mean(axis=0)
+                                    ) / (n_samples[i][ts-j]+1)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: Why are you doing the multiplication and division with the number of samples? I'm not familiar with this. Normally for the MSD is an average where you divide by the number of samples at the end.

File: package/MDAnalysis/lib/distances.py (hunk 15 above)
Comment (richardjgowers) [inline package/MDAnalysis/lib/distances.py]: I've got a feeling that he might have forked from master not develop?  These seem like regressions to non numpy docstrings

Comment (abiedermann) [inline package/MDAnalysis/lib/distances.py]: Yes, unfortunately I think I did that.

File: package/MDAnalysis/analysis/MSD.py (hunk 9 above)
Comment (abiedermann) [inline package/MDAnalysis/analysis/MSD.py]: I should be able to modify the state_dynamics script to extend AnalysisBase, but it would be difficult to use that frame work for this script.

**Comment by abiedermann** [comment]
I removed the state_dynamics.py module from the develop branch, and will submit it as a separate pull request for it after fixing documentation, nomenclature, and extending AnalysisBase.

File: package/MDAnalysis/analysis/MSD.py (hunk 9 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This script can also be modified to use AnalysisBase but that change isn't straight forward. You can keep it as is right now but we should try to make this look like it was derived from AnalysisBase.

**Comment by kain88-de** [comment]
@MDAnalysis/coredevs. Please don't merge this PR via github. Rather use the CLI interface and remove the large 14MB test files from the history, this might not be straight forward due to the merge-commit included in the PR. This will keep our repository smaller when people download it.

File: package/CHANGELOG [hunk 17]
```diff
@@ -14,12 +14,16 @@ The rules for this file:
...
 ------------------------------------------------------------------------------
 ??/??/16 kain88-de, fiona-naughton, richardjgowers, tyler.je.reddy, jdetle
-         euhruska, orbeckst
+         euhruska, orbeckst, abiedermann
 
   * 0.15.1
 
 Enhancements
 
+  * Added state_dynamics.py to MDAnalysis.analysis, enabling calculation
```
Comment (kain88-de) [inline package/CHANGELOG]: Please remove this as it will be added in a different PR

File: package/MDAnalysis/analysis/MSD.py [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+        selection X tau
+        (MSD is calculated as msd/n_samples)
+    n_samples : np.array(dtype=int)
+        Contains the number of samples corresponding to each msd
+        entry, organized:
+        selection X tau
+
+    Methods
+    -------
+    _select(self)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: You don't need to document private methods. Only the public methods to interact with the class.

File: package/MDAnalysis/analysis/MSD.py [hunk 19]
```diff
@@ -0,0 +1,398 @@
...
+            If true, include tau values after stop_frame in calculation (this
+            is useful for avoiding data loss when breaking large calculations
+            across multiple batch jobs)
+        """
+
+        self.universe = universe
+        self.select_list = select_list
+        if isinstance(select_list, str):
+            self.select_list = select_list.split(',')
+        print(self.select_list)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: We do not print values during normal execution. If you want to add verbosity we normally use a `quiet` kwarg as a switch.

File: package/MDAnalysis/analysis/MSD.py [hunk 20]
```diff
@@ -0,0 +1,398 @@
...
+        self.start_frame = int(start_frame)
+        self.stop_frame = int(stop_frame)
+        self.dt_restart = int(dt_restart)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: I would recommend to use `u.trajectory.check_slice_indices(start_frame,stop_frame, dt_restart)`. The function makes sure that the defined slice is valid and returns integers

Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This will also mean that you need to introduce a `step` argument. Your `dt_restart` isn't really a step. But this is important to downsample a trajectory. Eg you sampled originally at 1ps but want to run the analysis for frames saved every 5 ps.

File: package/MDAnalysis/analysis/MSD.py [hunk 21]
```diff
@@ -0,0 +1,398 @@
...
+        self.out_name = out_name
+        # handling short simulation cases
+        n_frames = int(stop_frame) - int(start_frame) + 1
+        if int(len_msd) > int(n_frames):
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: In this case you should issue a warning about the algorithm stability. In my experience a correlation function is only reliable for about 10% of the total data. But completely random for values larger then half the length. Because then you only use one value to estimate the correlation value, a mean estimate from one value doesn't make a lot of sense.

File: package/MDAnalysis/analysis/MSD.py [hunk 22]
```diff
@@ -0,0 +1,398 @@
...
+        # initializing object attributes
+        self.pos, self.atomids_per_sel_at_t, self.map_atomids_to_pos_at_t, = \
+            self._init_pos_deque()
+
+        n_sel = len(self.pos)
+        self.msd = np.zeros((n_sel, self.len_msd), dtype='float64')
+        self.n_samples = np.zeros((n_sel, self.len_msd), dtype='int')
+
+        # check that trajectory length is correct
+        try:
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: check_slice_indices will to this check for you.

File: package/MDAnalysis/analysis/MSD.py [hunk 23]
```diff
@@ -0,0 +1,398 @@
...
+            assert len(self.universe.trajectory) >= \
+                    (self.stop_frame-self.start_frame+1)
+        except RuntimeError:
+            print("Sample interval exceeds trajectory length. This may result"
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: Please raise Exceptions to stop execution. But as mentioned above another convenience function can do this check for you.

File: package/MDAnalysis/analysis/MSD.py [hunk 24]
```diff
@@ -0,0 +1,398 @@
...
+        selections : list
+            list of atom groups
+        """
+        selections = []
+        for i, sel in enumerate(self.select_list):
+            selections.append(self.universe.select_atoms(sel))
+            if selections[i].n_atoms == 0:
+                selections.pop(i)
+                selections.append(None)
+
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This can be optimized even further. Because python uses a pbc condition for array indices you can use `[-1]` to access the last element. Also the pop/append is unnecessary if we just set the value directly.

```python
for sel in self.select_list:
    selections.append(self.universe.select_atoms(sel))
    if selections[-1].n_atoms == 0:
        selections[-1] = None
```

File: package/MDAnalysis/analysis/MSD.py [hunk 25]
```diff
@@ -0,0 +1,398 @@
...
+        """
+
+        n_sel = len(self.select_list)
+        
+        # dictionaries with key: (id) and value: (array index)
+        # selection X frame X dictionary
+        map_atomids_to_pos_at_t = [[dict() for j in range(self.len_msd)]
+                     for i in range(n_sel)]
+        # atom ids which satisfy the selection at each frame
+        atomids_per_sel_at_t = [[set() for j in range(self.len_msd)]
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: thanks that helps a lot with understanding the algorithm.

File: package/MDAnalysis/analysis/MSD.py [hunk 26]
```diff
@@ -0,0 +1,398 @@
...
+            if frame % (self.len_msd/10) == 0:
+                print("Frame: "+str(frame))
+        # converting lists to deques
+        for i in range(n_sel):
+            map_atomids_to_pos_at_t[i] = deque(map_atomids_to_pos_at_t[i],
+                                               maxlen=self.len_msd)
+            atomids_per_sel_at_t[i] = deque(atomids_per_sel_at_t[i],
+                                            maxlen=self.len_msd)
+            pos[i] = deque(pos[i], maxlen=self.len_msd)
+        print("Complete!")
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: see previous comment about verbosity.

File: package/MDAnalysis/analysis/MSD.py [hunk 27]
```diff
@@ -0,0 +1,398 @@
...
+        """Calls _process_pos_data and saves final output"""
+
+        self._process_pos_data()
+
+        if self.write_output:
+            pickle.dump([self.msd/self.n_samples, self.n_samples],
+                        open(self.out_name, 'wb'))
+        # for testing purposes
+        else:
+            return [self.msd/self.n_samples, self.n_samples]
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: To emulate the AnalysisBase API this function should always return `self` and not write any data. Writing data should be a separate function call.

File: package/MDAnalysis/analysis/MSD.py [hunk 28]
```diff
@@ -0,0 +1,398 @@
...
+        return
+
+    def run(self):
+        """Calls _process_pos_data and saves final output"""
+
+        self._process_pos_data()
+
+        if self.write_output:
+            pickle.dump([self.msd/self.n_samples, self.n_samples],
+                        open(self.out_name, 'wb'))
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: I would prefer a CSV output to allow further processing by different tools.

File: package/MDAnalysis/analysis/distances.py [hunk 29]
```diff
@@ -113,6 +112,56 @@ def contact_matrix(coord, cutoff=15.0, returntype="numpy", box=None):
             contact_matrix_no_pbc(coord, sparse_contacts, cutoff)
         return sparse_contacts
 
+def binary_contact_matrix(ref_coord, sel_coord, cutoff1, cutoff2, box):
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/distances.py]: Is this still needed for the MSD analysis?

File: package/MDAnalysis/analysis/MSD.py (hunk 19 above)
Comment (richardjgowers) [inline package/MDAnalysis/analysis/MSD.py]: Most of these print statements would make good `logger.debug` calls instead

https://fangpenlin.com/posts/2012/08/26/good-logging-practice-in-python/

**Comment by richardjgowers** [comment]
@kain88-de if this is squashed locally into a single commit, will it squash the large file out of existence? (I'm thinking yes?)

**Comment by kain88-de** [comment]
> @kain88-de if this is squashed locally into a single commit, will it squash the large file out of existence? (I'm thinking yes?)

Yes it would do that. But you can't just squash this PR due to the merge commits included.

**Comment by orbeckst** [comment]
> On 21 Oct, 2016, at 07:53, Richard Gowers <notifications@github.com> wrote:
> 
> @kain88-de if this is squashed locally into a single commit, will it squash the large file out of existence? (I'm thinking yes?)

It should.

--
Oliver Beckstein * orbeckst@gmx.net
skype: orbeckst  * orbeckst@gmail.com

**Comment by abiedermann** [comment]
I've implemented most of the changes we discussed, but still need to adjust the code to fit the AnalysisBase framework. I will add the step parameter while making this change. I will then double check the documentation and reformat the output csv file.

File: package/MDAnalysis/analysis/MSD.py [hunk 30]
```diff
@@ -63,8 +63,8 @@
 regardless of complexity, so the MSD of something like
  'type OT around 4 type HT' can be calculated.
 
-By default, the MSD object will write output in binary to a pickle file named
-'msd' which will contain MSD calculated for tau less than 250 frames or the
+By default, the MSD object will write output in binary to a csv file named
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: csv files aren't written in binary.

File: package/MDAnalysis/analysis/MSD.py [hunk 31]
```diff
@@ -128,22 +129,14 @@ class MSD(object):
...
     run(self):
         Calls _process_pos_data and saves final output
 
     """
 
     def __init__(self, universe, select_list, start_frame, stop_frame,
-                 dt_restart, out_name='msd', len_msd=250, write_output=True,
-                 max_data=False):
+                 dt_restart, out_name='msd.csv', len_msd=250, write_output=True,
+                 max_data=False,verbose=False):
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: In MDAnalysis we use `quiet` instead of `verbose`. Until we changed that (see #903) please use quiet

File: package/MDAnalysis/analysis/MSD.py [hunk 32]
```diff
@@ -168,24 +161,45 @@ def __init__(self, universe, select_list, start_frame, stop_frame,
...
+
+        self.start_frame, self.stop_frame, self.dt_restart = \
+            self.universe.trajectory.check_slice_indices(
+                          start_frame,stop_frame,dt_restart)
+
         self.out_name = out_name
+
+        # initializing logger object
+        logging.basicConfig(level=logging.INFO if verbose
+                            else logging.WARNING)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: If you are quiet/not verbose that don't print/log anything.

File: package/MDAnalysis/analysis/MSD.py [hunk 33]
```diff
@@ -168,24 +161,45 @@ def __init__(self, universe, select_list, start_frame, stop_frame,
...
+        self.logger = logging.getLogger(__name__)
+
+        self.logger.info(self.select_list)
+        
         # handling short simulation cases
         n_frames = int(stop_frame) - int(start_frame) + 1
         if int(len_msd) > int(n_frames):
             len_msd = int(n_frames)
+            self.logger.warning(
+                "Correcting len_msd to the number for frames in the trajectory. ")
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/MSD.py]: This should raise an exception instead of silently changing the input. Not everyone looks into the logs especially in an interactive session. It might take the user quite some time to figure out that is trajectory isn't long enough, since the code he uses looks correct and no visible warning is printed.

**Comment by abiedermann** [comment]
Sorry for the delay in responding. Based on my initial results obtained using this MSD code, it looks like I am going to need to make some further modifications to the code for my research. I am going to decouple the trajectory reading and calculation steps, so that the code generates a larger intermediate data structure and then performs a fast fourier transform to calculate MSD. While this has more memory overhead, this approach will get rid of the msd_len parameter setting which may discard useful parts of the MSD vs tau plot.
This is a more substantial revision, so it may be easier to submit a new pull request taking into account all the things we discussed from this one and fixing the branching issue. I am focused on grad school apps right now, but I should be able to complete these changes by early January.

**Comment by richardjgowers** [comment]
@abiedermann no problem.  We've made some pretty big changes to the develop branch recently, so it might be easier to rebranch off that too.  I'm going to close this, but feel free to open a new PR and/or ask questions on the [dev board](https://groups.google.com/forum/#!forum/mdnalysis-devel)
